CRAMcodecs.tex

```diff
@@ -0,0 +1,3183 @@
+  \State $value \gets 0$
+  \State $length \gets 0$
+  \Repeat
+    \State $c \gets$ \Call{ReadUint8}{}
+    \State $value \gets (value \shiftl 7) + (c \bitand 127)$
+    \State $length \gets length + 1$
+  \Until{$c < 128$}
+  \State \Return ($value$, $length$) \Comment{or just $value$ if caller uses only that}
+  \EndFunction
+\end{algorithmic}
+
+\section{rANS 4x8 - Asymmetric Numeral System}
+
+% Lifted over from CRAMv3.tex
+
+This is the rANS format first defined in CRAM v3.0.
+
+rANS is the range-coder variant of the Asymmetric Numerical
+System\footnote{J. Duda, \textit{Asymmetric numeral systems: entropy
+    coding combining speed of Huffman coding with compression rate of
+    arithmetic coding}, \url{http://arxiv.org/abs/1311.2540}}.
+
+The structure of the external rANS codec consists of several
+components: meta-data consisting of compression-order, and compressed
+and uncompressed sizes; normalised frequencies of the alphabet systems
+to be encoded, either in Order-0 or Order-1 context; and the rANS
+encoded byte stream itself.
+
+Here "Order" refers to the number of bytes of context used in
+computing the frequencies. It will be 0 or 1.  Ignoring punctuation
+and space, an Order-0 analysis of English text may observe that `e' is
+the most common letter (12-13\%), and that `u' occurs only around 2.5\%
+of the time.  If instead we consider the frequency of a letter in the
+context of one previous letter (Order-1) then these statistics change
+considerably;  we know that if the previous letter was `q' then `e'
+becomes a rare letter while `u' is the most likely.
+
+These observed frequencies are directly related to the amount of
+storage required to encode a symbol (e.g. an alphabet
+letter)\footnote{ C.E. Shannon, \textit{A Mathematical Theory of
+    Communication}, Bell System Technical Journal, vol. 27,
+    pp. 379-423, 623-656, July, October, 1948}.
+
+
+\subsubsection*{\textbf{rANS 4x8 compressed data structure}}
+A compressed data block consists of the following logical parts:
+
+\begin{tabular}{|l|l|>{\raggedright}p{100pt}|>{\raggedright}p{220pt}|}
+\hline
+\textbf{Value data type} & \textbf{Name} & \textbf{Description}\tabularnewline
+\hline
+byte & order & the order of the codec, either 0 or 1\tabularnewline
+\hline
+int & compressed size & the size in bytes of frequency table and compressed blob\tabularnewline
+\hline
+int & data size & raw or uncompressed data size in bytes\tabularnewline
+\hline
+byte[] & frequency table & byte frequencies of input data written using RLE\tabularnewline
+\hline
+byte[] & compressed blob & compressed data\tabularnewline
+\hline
+\end{tabular}
+
+\subsection{\textbf{Frequency table}}
+
+The alphabet used here is simply byte values, so a maximum of 256
+symbols as some values may not be present.
+
+The symbol frequency table indicates which symbols are present and
+what their relative frequencies are.  The total sum of symbol
+frequencies are normalised to add up to 4095.
```

Comment:
Both htscodecs and HTSJDK add up to 4096, not 4095.

* https://github.com/samtools/htscodecs/blob/843d4f63b1c64905881b4648916a4d027baa1a1c/htscodecs/rANS_static.c#L37-L38
* https://github.com/samtools/htsjdk/blob/4a4024a97ee3e87096df6ad9b22c8260bd527772/src/main/java/htsjdk/samtools/cram/compression/rans/Constants.java#L4-L5

Reply:
I think this is a legacy thing from CRAM 3.0 which wrongly stated 4095 as I believed 4096 would cause wrap-around issues.  That wasn't true, and the code works just fine.  I'm a little cautious about explicitly permitting 4096 incase people have implemented a decoder with an assumption that it matches what the spec originally stated, but I think we can enumerate the implementations pretty easily to double check this.

I'll investigate.  It would definitely be cleaner to recommend 4096 (with a comment that 4095 is permitted for compatibility with historic versions of the spec).

Reply:
I can verify that the output from Scramble (io_lib's original CRAM implementation), htslib, and htsjdk all produce sums of 4095 for their O0 and O1 frequency tables.

The TOTFREQ definition in Java just copied my C version, so it all stems from the same silly issue back then I guess.  4096 is the size of the array used in the decode function, and it's that simply because it permits AND operations and lookup tables (which is probably why I originally confused myself with 4095 being the maximum freq).  The renormalisation of frequencies is non-trivial and isn't covered by the specification either as it describes the at-rest file format rather than a specific implementation for determining how to get your frequencies summing correctly.  So it's inevitable that we start with a target value and then have to rescale after summing to get things adding up.

I agree it looks like the javascript implementation is wrong here.  It needs fixing.